sympy/polys/domainmatrix.py

```diff
@@ -0,0 +1,653 @@
+    """
+    def __init__(self, rowslist, shape, domain):
+        super().__init__(rowslist)
+        self.shape = self.rows, self.cols = m, n = shape
+        self.domain = domain
+
+        if not (len(self) == m and all(len(row) == n for row in self)):
+            raise DDMBadInputError("Inconsistent row-list/shape")
+
+    def __str__(self):
+        cls = type(self).__name__
+        rows = list.__str__(self)
+        return '%s(%s, %s, %s)' % (cls, rows, self.shape, self.domain)
+
+    def __eq__(self, other):
+        if not isinstance(other, DDM):
+            return False
+        return (super().__eq__(other) and self.domain == other.domain)
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    @classmethod
+    def zeros(cls, shape, domain):
+        z = domain.zero
+        m, n = shape
+        rowslist = ([z] * n for _ in range(m))
+        return DDM(rowslist, shape, domain)
+
+    @classmethod
+    def eye(cls, size, domain):
+        one = domain.one
+        ddm = cls.zeros((size, size), domain)
+        for i in range(size):
+            ddm[i][i] = one
+        return ddm
+
+    def copy(self):
+        copyrows = (row[:] for row in self)
+        return DDM(copyrows, self.shape, self.domain)
+
+    def __add__(a, b):
+        if not isinstance(b, DDM):
+            return NotImplemented
+        return a.add(b)
+
+    def __sub__(a, b):
+        if not isinstance(b, DDM):
+            return NotImplemented
+        return a.sub(b)
+
+    def __neg__(a):
+        return a.neg()
+
+    def __mul__(a, b):
+        if b in a.domain:
+            return a.mul(b)
+        else:
+            return NotImplemented
+
+    def __matmul__(a, b):
+        if isinstance(b, DDM):
+            return a.matmul(b)
+        else:
+            return NotImplemented
+
+    @classmethod
+    def _check(cls, a, op, b, ashape, bshape):
+        if a.domain != b.domain:
+            msg = "Domain mismatch: %s %s %s" % (a.domain, op, b.domain)
+            raise DDMDomainError(msg)
+        if ashape != bshape:
+            msg = "Shape mismatch: %s %s %s" % (a.shape, op, b.shape)
+            raise DDMShapeError(msg)
+
+    def add(a, b):
+        """a + b"""
+        a._check(a, '+', b, a.shape, b.shape)
+        c = a.copy()
+        ddm_iadd(c, b)
+        return c
+
+    def sub(a, b):
+        """a - b"""
+        a._check(a, '-', b, a.shape, b.shape)
+        c = a.copy()
+        ddm_isub(c, b)
+        return c
+
+    def neg(a):
+        """-a"""
+        b = a.copy()
+        ddm_ineg(b)
+        return b
+
+    def matmul(a, b):
+        """a @ b (matrix product)"""
+        m, o = a.shape
+        o2, n = b.shape
+        a._check(a, '*', b, o, o2)
+        c = a.zeros((m, n), a.domain)
+        ddm_imatmul(c, a, b)
+        return c
+
+    def rref(a):
+        """Reduced-row echelon form of a and list of pivots"""
+        b = a.copy()
+        pivots = ddm_irref(b)
+        return b, pivots
```

Comment:
It was discussed back in
https://github.com/sympy/sympy/issues/18445#issuecomment-578549327
that whether pivots should be returned with the matrix or not.
However, since this classes are for much more lower level implementation, this can be different.

Reply:
Yes, I remembered that. I think at this lower-level we should just organise things around what is efficient. Computing the pivots is effectively free so we might as well return them